agent/consul/grpc_resolver.go

```diff
@@ -0,0 +1,105 @@
+package consul
+
+import (
+	"sync"
+
+	"github.com/hashicorp/consul/agent/metadata"
+	"google.golang.org/grpc/resolver"
+)
+
+const grpcResolverScheme = "consul"
+
+var resolverBuilder *ServerResolverBuilder
+
+func init() {
+	resolverBuilder = NewServerResolverBuilder()
+	resolver.Register(resolverBuilder)
+}
+
+// ServerResolverBuilder tracks the current server list and keeps any
+// ServerResolvers updated when changes occur.
+type ServerResolverBuilder struct {
+	servers   map[string]*metadata.Server
+	resolvers map[*ServerResolver]struct{}
```

Comment:
Why are there multiple resolvers? What is the significance of separate ones?

Reply:
Each call to grpc.Dial gets a new ClientConn which comes with its own resolver. The ClientConn can be re-used for multiple rpc calls and rebalance itself to a healthy server under the hood. I explain in a comment lower down why it ended up being a map instead of one ClientConn for everything